kernel/kernel-api/src/main/java/io/delta/kernel/internal/rowtracking/RowTracking.java

```diff
@@ -0,0 +1,155 @@
+   * committing. If row tracking is not supported by the given protocol, returns the input actions
+   * unchanged.
+   *
+   * <p>When needed, it sets each {@link AddFile} action’s base row ID based on the current high
+   * watermark and increments the watermark accordingly. If a default row commit version is missing,
+   * it assigns the provided commit version.
+   *
+   * @param snapshot the current snapshot of the table
+   * @param commitVersion the version of the commit for default row commit version assignment
+   * @param dataActions the {@link CloseableIterable} of data actions to process
+   * @return an {@link CloseableIterable} of data actions with base row IDs and default row commit
+   *     versions assigned
+   */
+  public static CloseableIterable<Row> assignBaseRowIdAndDefaultRowCommitVersion(
+      SnapshotImpl snapshot, long commitVersion, CloseableIterable<Row> dataActions) {
+    checkArgument(
+        TableFeatures.isRowTrackingSupported(snapshot.getProtocol()),
+        "Base row ID and default row commit version are assigned "
+            + "only when feature 'rowTracking' is supported.");
+
+    return new CloseableIterable<Row>() {
+      @Override
+      public void close() throws IOException {
+        dataActions.close();
+      }
+
+      @Override
+      public CloseableIterator<Row> iterator() {
+        // Used to keep track of the current high watermark as we iterate through the data actions.
+        // Use an AtomicLong to allow for updating the high watermark in the lambda.
+        final AtomicLong currRowIdHighWatermark = new AtomicLong(readRowIdHighWaterMark(snapshot));
+        return dataActions
+            .iterator()
+            .map(
+                row -> {
+                  // Non-AddFile actions are returned unchanged
+                  if (row.isNullAt(ADD_FILE_ORDINAL)) {
+                    return row;
+                  }
+
+                  AddFile addFile = AddFile.fromRow(row.getStruct(ADD_FILE_ORDINAL));
+
+                  // Assign base row ID if missing
+                  if (!addFile.getBaseRowId().isPresent()) {
+                    final long numRecords = getNumRecordsOrThrow(addFile);
+                    addFile = addFile.withNewBaseRowId(currRowIdHighWatermark.get() + 1L);
+                    currRowIdHighWatermark.addAndGet(numRecords);
+                  }
+
+                  // Assign default row commit version if missing
+                  if (!addFile.getDefaultRowCommitVersion().isPresent()) {
+                    addFile = addFile.withNewDefaultRowCommitVersion(commitVersion);
+                  }
+
+                  // Return a new AddFile row with assigned baseRowId/defaultRowCommitVersion
+                  return SingleAction.createAddFileSingleAction(addFile.toRow());
+                });
+      }
+    };
+  }
+
+  /**
+   * Emits a {@link DomainMetadata} action if the row ID high watermark has changed due to newly
```

Comment:
`Emits` -> `Returns`

Mention that: This method requires that rowTracking feature is supported